app/shop.py

```diff
@@ -0,0 +1,33 @@
+import datetime
+
+
+class Shop:
+    def __init__(self, name: str, location: list[int], products: dict) -> None:
+        self.name = name
+        self.location = location
+        self.products = products
+
+    def calculate_product_cost(self, product_cart: dict) -> float:
+        total_cost = 0
+
+        for product, amount in product_cart.items():
+            if product in self.products:
+                total_cost += self.products[product] * amount
+        return total_cost
+
+    def print_receipt(self, customer_name: str, product_cart: dict) -> None:
+        print(
+            f'\nDate: {datetime.datetime.now().strftime("%d/%m/%Y %H:%M:%S")}'
+        )
+        print(f"Thanks, {customer_name}, for your purchase!")
+        print("You have bought:")
+
+        for product, amount in product_cart.items():
+            cost = self.products[product] * amount
+            formatted_cost = (f"{round(cost, 1)}" if cost % 1 != 0
+                              else f"{int(cost)}")
+            print(f"{amount} {product}s for {formatted_cost} dollars")
```

Comment:
The code for calculating and formatting the cost of each product is duplicated in the `print_receipt` method and the `calculate_product_cost` method. Consider creating a separate method to handle this calculation to adhere to the DRY principle.